### DISCUSSION TITLE
命令行执行 vs gradio界面执行

### DISCUSSION THREAD

**Discussion by bbeyondllove** [modelscope/FunClip#58] [Q&A] [answered]
python paraclipper/videoclipper.py --stage 1 --file  bk_man_1.wav  --output_dir ./output
这个命令执行大概要15秒左右，音频才3秒。不过在界面执行是很快的，找不到问题在哪。

**Comment by R1ckShi** [comment, marked as answer]
命令行调用在stage1需要load模型，通过gradio调用是模型的加载相关操作已经在启动服务时完成了。

**Comment by bbeyondllove** [comment]
> 命令行调用在stage1需要load模型，通过gradio调用是模型的加载相关操作已经在启动服务时完成了。

嗯，那是不是可以做成http api，先加载模型，然后命令行相当于client来调用，这样应该会快很多

**Comment by R1ckShi** [comment]
是可以的，但是我觉得这个应用的场景很少有批量自动调用的，都需要结合每一条视频的识别结果去选择需要的段落。所以主要推的是gradio界面的调用方式。

**Comment by LauraGPT** [comment]
上面的维护者回答仍然准确：这里慢的主要不是“3 秒音频的纯推理”，而是 **CLI 进程冷启动时重新加载模型**。

当前主干仍然是这个结构：

- `videoclipper.py --stage 1` 每启动一次进程，都会在 `runner()` 内重新构造 ASR、VAD、标点和说话人模型，然后才读取音频并识别（[源码](https://github.com/modelscope/FunClip/blob/80ea6ea18b8937db0d453e3f5440fdfbc4778a32/funclip/videoclipper.py#L445-L486)）。因此短音频的总耗时很容易被模型加载时间主导。
- Gradio 服务则在进程启动时创建一次 `AutoModel` 和 `VideoClipper`，后续请求复用同一个模型实例（[源码](https://github.com/modelscope/FunClip/blob/80ea6ea18b8937db0d453e3f5440fdfbc4778a32/funclip/launch.py#L55-L70)）。

所以，如果需要批量或程序化调用，不要为每个文件执行一次 `videoclipper.py --stage 1`，而应启动常驻服务：

```bash
python funclip/launch.py --listen --port 7860
```

这也是当前 README 给出的[本地 Gradio 服务方式](https://github.com/modelscope/FunClip/blob/80ea6ea18b8937db0d453e3f5440fdfbc4778a32/README.md#L107-L128)。Gradio 本身就是 HTTP 服务；客户端建议使用 `gradio_client`，因为它会正确处理文件上传、队列和会话状态：

```python
from gradio_client import Client

client = Client("http://127.0.0.1:7860")
print(client.view_api())
```

本地完整 FunClip 的 API 名称和参数应以 `view_api()` 的实时输出为准，不要把某个 Gradio 版本自动生成的路径硬编码进长期客户端。需要连续“识别后再剪辑”时，也要复用同一个 `Client`，因为中间的时间戳/句子状态属于该会话。

### 当前公开 API 的可运行示例

官方 [FunClip Hugging Face Space](https://huggingface.co/spaces/FunAudioLLM/FunClip) 当前公开了三个固定端点，可以直接验证常驻模型的调用方式：

```bash
pip install gradio_client
```

```python
from gradio_client import Client, handle_file

client = Client("FunAudioLLM/FunClip")
video = {
    "video": handle_file("input.mp4"),
    "subtitles": None,
}

text = client.predict(video, api_name="/transcribe_video")
selector = client.predict(api_name="/build_selector")

# 选择前两个片段；继续复用同一个 client。
selected = [item[1] for item in selector["choices"][:2]]
clip, info = client.predict(
    video,
    selected,
    api_name="/clip_video",
)
```

我在当前 Space revision `e1344660`、`gradio_client==1.14.0` 上用官方 33.833 秒示例视频实测：

```text
/transcribe_video   8.566 s，返回 13 个带时间戳句段
/build_selector     2.178 s，取回同一会话的 13 个选项
/clip_video         7.162 s，前两个相邻片段合并为 2.640 s 视频
输出文件            125,356 bytes，FFprobe 可正常读取
```

公开 Space 使用共享的 CPU 实例，因此这些时间包含上传、排队和服务负载，**不能当作本地模型 benchmark**；这次实测只用于确认 API、会话复用和剪辑链路当前确实可用。该轻量 Space 目前只接收视频并使用中文 Paraformer；音频输入、多模型选择和完整 FunClip 功能请启动上面的本地服务。公开 Space 的实现也能看到模型是在模块加载阶段创建一次，而不是每个请求重新加载（[源码](https://huggingface.co/spaces/FunAudioLLM/FunClip/blob/e13446600e24ca999f82f7720e6ad8e1610edbbe/app.py)）。

一句话总结：**CLI 单次命令适合偶发任务；批量任务应让模型常驻，并通过同一个 Gradio/HTTP 客户端连续调用。**
